src/main/drivers/max7456.c

```diff
@@ -327,110 +428,135 @@ void max7456Write(uint8_t x, uint8_t y, const char *buff, uint8_t mode)
+    int bufPtr = 0;
     int pos;
     uint_fast16_t updatedCharCount;
-    uint8_t currentMode;
+    uint8_t charMode;
 
-    // Check if device is available
-    if (max7456dev == NULL) {
-        return;
-    }
+    for (pos = 0, updatedCharCount = 0;;) {
+        pos = BITARRAY_FIND_FIRST_SET(screenIsDirty, pos);
+        if (pos < 0 || pos >= maxScreenSize) {
+            // No more dirty chars.
+            break;
+        }
 
-    if (!max7456Lock && !fontIsLoading) {
-        // (Re)Initialize MAX7456 at startup or stall is detected.
+        // Found one dirty character to send
+        uint8_t ph = pos >> 8;
+        uint8_t pl = pos & 0xff;
 
-        max7456Lock = true;
-
-        // Stall check
-        busRead(max7456dev, MAX7456ADD_VM0 | MAX7456ADD_READ, &stallCheck);
-
-        nowMs = millis();
-
-        if (stallCheck != videoSignalReg) {
-            max7456ReInit();
-        } else if ((videoSignalCfg == VIDEO_SYSTEM_AUTO)
-                  && ((nowMs - lastSigCheckMs) > MAX7456_SIGNAL_CHECK_INTERVAL_MS)) {
-
-            // Adjust output format based on the current input format.
-            busRead(max7456dev, MAX7456ADD_STAT, &videoSense);
-
-            if (videoSense & STAT_LOS) {
-                videoDetectTimeMs = 0;
-            } else {
-                if ((VIN_IS_PAL(videoSense) && VIDEO_MODE_IS_NTSC(videoSignalReg))
-                  || (VIN_IS_NTSC_alt(videoSense) && VIDEO_MODE_IS_PAL(videoSignalReg))) {
-                    if (videoDetectTimeMs) {
-                        if (millis() - videoDetectTimeMs > VIDEO_SIGNAL_DEBOUNCE_MS) {
-                            max7456ReInit();
-                        }
-                    } else {
-                        // Wait for signal to stabilize
-                        videoDetectTimeMs = millis();
-                    }
-                }
+        charMode = MODE_BYTE(screenBuffer[pos]);
+        uint8_t chr = CHAR_BYTE(screenBuffer[pos]);
+        if (CHAR_MODE_IS_EXT(charMode)) {
+            if (!DMM_IS_8BIT_MODE(state.registers.dmm)) {
+                state.registers.dmm |= DMM_8BIT_MODE;
+                bufPtr = max7456PrepareBuffer(spiBuff, bufPtr, MAX7456ADD_DMM, state.registers.dmm);
             }
 
-            lastSigCheckMs = nowMs;
-        }
-
-        //------------   end of (re)init-------------------------------------
-        uint8_t spiBuff[MAX_CHARS2UPDATE * BYTES_PER_CHAR2UPDATE];
-        int bufPtr = 0;
-
-        for (pos = 0, updatedCharCount = 0;;) {
-            pos = BITARRAY_FIND_FIRST_SET(screenIsDirty, pos);
-            if (pos < 0 || pos >= maxScreenSize) {
-                // No more dirty chars.
-                break;
-            }
+            bufPtr = max7456PrepareBuffer(spiBuff, bufPtr, MAX7456ADD_DMAH, ph | DMAH_8_BIT_DMDI_IS_CHAR_ATTR);
+            bufPtr = max7456PrepareBuffer(spiBuff, bufPtr, MAX7456ADD_DMAL, pl);
+            bufPtr = max7456PrepareBuffer(spiBuff, bufPtr, MAX7456ADD_DMDI, charMode << 2);
```

Comment:
It's not obvious that `charMode << 2` is actually referring to extra bit of 16-bit character. We should probably need to comment on this.

Reply:
It definitely needs a comment, but the shift not because of the extra bit for 9 bit chars. The reason is that we're storing that byte in `DMM` format, but when writing attributes via `DMDI` the bits are 2 positions up.

Reply:
So you see - I missed that point 😄